src/cloud-api-adaptor/pkg/userdata/provision.go

```diff
@@ -149,60 +164,82 @@ func parseUserData(userData []byte) (*CloudConfig, error) {
 	return &cc, nil
 }
 
-func findConfigEntry(path string, cc *CloudConfig) []byte {
-	for _, wf := range cc.WriteFiles {
-		if wf.Path != path {
-			continue
-		}
-		return []byte(wf.Content)
+func writeFile(path string, bytes []byte) error {
+	// Ensure the parent directory exists
+	err := os.MkdirAll(filepath.Dir(path), 0755)
+	if err != nil {
+		return fmt.Errorf("failed to create directory: %w", err)
+	}
+
+	err = os.WriteFile(path, bytes, 0644)
+	if err != nil {
+		return fmt.Errorf("failed to write file %s: %w", path, err)
 	}
+	logger.Printf("Wrote %s\n", path)
 	return nil
 }
 
-type entry struct {
-	path     string
-	optional bool
+func processCloudConfig(cfg *Config, cc *CloudConfig) error {
+	for _, wf := range cc.WriteFiles {
+		path := wf.Path
+		bytes := []byte(wf.Content)
+		if bytes != nil {
+			if err := writeFile(path, bytes); err != nil {
+				return fmt.Errorf("failed to write config file %s: %w", path, err)
+			}
+		}
+	}
+
+	return nil
 }
 
-func (f *entry) writeFile(cc *CloudConfig) error {
-	bytes := findConfigEntry(f.path, cc)
-	if bytes == nil {
-		if !f.optional {
-			return fmt.Errorf("failed to find %s entry in cloud config", f.path)
-		}
-		return nil
+func extractInitdataAndHash(cfg *Config) error {
+	if _, err := os.Stat(cfg.initdataPath); err != nil {
+		return fmt.Errorf("Error stat initdata file: %w", err)
 	}
 
-	// Ensure the parent directory exists
-	err := os.MkdirAll(filepath.Dir(f.path), 0755)
+	dataBytes, err := os.ReadFile(cfg.initdataPath)
 	if err != nil {
-		return fmt.Errorf("failed to create directory: %w", err)
+		return fmt.Errorf("Error read initdata file: %w", err)
 	}
 
-	err = os.WriteFile(f.path, bytes, 0644)
+	decodedBytes, err := base64.StdEncoding.DecodeString(string(dataBytes))
 	if err != nil {
-		return fmt.Errorf("failed to write file: %w", err)
+		return fmt.Errorf("Error base64 decode initdata: %w", err)
 	}
-	logger.Printf("Wrote %s\n", f.path)
-	return nil
-}
-
-func processCloudConfig(cfg *Config, cc *CloudConfig) error {
-	entries := []entry{
-		{path: cfg.paths.agentConfig, optional: false},
-		{path: cfg.paths.daemonConfig, optional: false},
-		{path: cfg.paths.aaConfig, optional: true},
-		{path: cfg.paths.cdhConfig, optional: true},
-		{path: cfg.paths.authJson, optional: true},
+	initdata := InitData{}
+	err = toml.Unmarshal(decodedBytes, &initdata)
+	if err != nil {
+		return fmt.Errorf("Error unmarshalling initdata: %w", err)
 	}
 
-	for _, e := range entries {
-		err := e.writeFile(cc)
+	for key, value := range initdata.Data {
+		err := writeFile(filepath.Join(cfg.parentPath, key), []byte(value))
```

Comment:
what if `key` is `/run/coco/aa.toml`? I suppose it will fail, if `/run/coco` doesn't exist. Should we maybe specify a clear allowlist of valid `key` values and ignore all other keys? I think `StaticFiles` was supposed to be this, but it's not used as far as I can see?

Reply:
Yes, added allowlist for items can be in `initdata` and `write_files`.